File: packages/backend-canvas2d/src/picture-resources.ts
import { decodeDisplayList } from "./display-list";
import {
  ABI_VERSION,
  INSTRUCTION_FLAG_MASK,
  INSTRUCTION_FLAG_OPTIONAL,
  INSTRUCTION_HEADER_BYTES,
  INSTRUCTION_LENGTH_ESCAPE,
  MAX_PICTURE_RESOURCE_INSTRUCTIONS,
  MAX_PICTURE_RESIDENT_BYTES,
  MAX_PICTURE_RESOURCES_BYTES,
  MAX_RESOURCE_BYTES,
  MINIMUM_READABLE_ABI_VERSION,
  PICTURE_RESOURCE_LAYOUTS,
  PICTURE_RESOURCES_MAGIC,
  PROTOCOL_ALIGNMENT,
  PictureResourceOpcode,
  STREAM_HEADER_BYTES,
} from "./generated";

/** Fully validated immutable Picture lifecycle delta. */
export type PictureResourceDelta =
  | { readonly type: "define"; readonly pictureId: number; readonly bytes: Uint8Array }
  | { readonly type: "release"; readonly pictureId: number };

/** Deterministic failure raised before backend registry state changes. */
export class PictureResourceError extends Error {
  public constructor(message: string) {
    super(message);
    this.name = "PictureResourceError";
  }
}

/** Decodes a complete batch and recursively validates every nested DisplayList. */
export function decodePictureResourceBatch(input: Uint8Array): readonly PictureResourceDelta[] {
  if (input.byteLength > MAX_PICTURE_RESOURCES_BYTES) fail("picture batch exceeds maximum size");
  if (input.byteLength % PROTOCOL_ALIGNMENT !== 0) fail("picture batch is not aligned");
  const reader = new Reader(input);
  if (reader.u32() !== PICTURE_RESOURCES_MAGIC) fail("wrong picture-resource magic");
  if (reader.u16() < MINIMUM_READABLE_ABI_VERSION) fail("unsupported picture-resource ABI version");
  if (reader.u16() !== STREAM_HEADER_BYTES) fail("invalid picture-resource header length");
  if (reader.u32() !== input.byteLength) fail("picture-resource length does not match input");
  const declaredCount = reader.u32();
  if (declaredCount > MAX_PICTURE_RESOURCE_INSTRUCTIONS)
    fail("picture-resource instruction count exceeds limit");
  if (declaredCount > Math.floor(reader.remaining / 8))
    fail("picture-resource count cannot fit in the remaining bytes");

  const deltas: PictureResourceDelta[] = [];
  const seen = new Set<number>();
  let payloadTotal = 0;
  let actualCount = 0;
  while (reader.remaining > 0) {
    const start = reader.offset;
    const header = reader.instruction();
    actualCount += 1;
    if (!isKnownOpcode(PictureResourceOpcode, header.opcode)) {
      if (!header.optional) fail(`unknown picture-resource opcode ${String(header.opcode)}`);
      reader.seekTo(header.end);
      continue;
    }
    const opcode = header.opcode;
    const pictureId = nonzeroId(reader.u32());
    if (seen.has(pictureId)) fail("picture id occurs more than once in a batch");
    seen.add(pictureId);
    let delta: PictureResourceDelta;
    if (opcode === PictureResourceOpcode.DefinePicture) {
      const payloadBytes = reader.u32();
      if (payloadBytes > MAX_RESOURCE_BYTES) fail("picture payload exceeds resource limit");
      payloadTotal = checkedAdd(payloadTotal, payloadBytes);
      if (payloadTotal > MAX_PICTURE_RESIDENT_BYTES)
        fail("picture payloads exceed transaction budget");
      const bytes = reader.bytes(payloadBytes).slice();
      decodeDisplayList(bytes);
      const padding =
        (PROTOCOL_ALIGNMENT - (payloadBytes % PROTOCOL_ALIGNMENT)) % PROTOCOL_ALIGNMENT;
      for (const byte of reader.bytes(padding)) {
        if (byte !== 0) fail("picture payload padding must be zero");
      }
      delta = Object.freeze({ type: "define", pictureId, bytes });
    } else {
      delta = Object.freeze({ type: "release", pictureId });
    }
    const layout = PICTURE_RESOURCE_LAYOUTS[opcode];
    const consumed = reader.offset - start;
    if (
      consumed < layout.minimumBytes ||
      (layout.fixedBytes !== null && consumed !== layout.fixedBytes)
    ) {
      fail("picture-resource instruction length does not match schema");
    }
    if (reader.offset !== header.end) fail("instruction length does not match its payload");
    deltas.push(delta);
  }
  if (actualCount !== declaredCount) fail("picture-resource instruction count does not match");
  return Object.freeze(deltas);
}

/** Encodes canonical bytes for cross-language fixtures and contract tests. */
export function encodePictureResourceBatch(deltas: readonly PictureResourceDelta[]): Uint8Array {
  if (deltas.length > MAX_PICTURE_RESOURCE_INSTRUCTIONS)
    fail("picture-resource instruction count exceeds limit");
  const writer = new Writer();
  writer.u32(PICTURE_RESOURCES_MAGIC);
  writer.u16(ABI_VERSION);
  writer.u16(STREAM_HEADER_BYTES);
  writer.u32(0);
  writer.u32(deltas.length);
  const seen = new Set<number>();
  let payloadTotal = 0;
  for (const delta of deltas) {
    const pictureId = nonzeroId(delta.pictureId);
    if (seen.has(pictureId)) fail("picture id occurs more than once in a batch");
    seen.add(pictureId);
    writer.instruction(
      delta.type === "define"
        ? PictureResourceOpcode.DefinePicture
        : PictureResourceOpcode.ReleasePicture,
    );
    writer.u32(pictureId);
    if (delta.type === "define") {
      decodeDisplayList(delta.bytes);
      if (delta.bytes.byteLength > MAX_RESOURCE_BYTES)
        fail("picture payload exceeds resource limit");
      payloadTotal = checkedAdd(payloadTotal, delta.bytes.byteLength);
      if (payloadTotal > MAX_PICTURE_RESIDENT_BYTES)
        fail("picture payloads exceed transaction budget");
      writer.u32(delta.bytes.byteLength);
      writer.bytes(delta.bytes);
      writer.pad();
    }
  }
  const output = writer.finish();
  if (output.byteLength > MAX_PICTURE_RESOURCES_BYTES) fail("picture batch exceeds maximum size");
  new DataView(output.buffer, output.byteOffset, output.byteLength).setUint32(
    8,
    output.byteLength,
    true,
  );
  return output;
}

class Reader {
  readonly #input: Uint8Array;
  readonly #view: DataView;
  #offset = 0;

  public constructor(input: Uint8Array) {
    this.#input = input;
    this.#view = new DataView(input.buffer, input.byteOffset, input.byteLength);
  }

  public get offset(): number {
    return this.#offset;
  }
  public get remaining(): number {
    return this.#input.byteLength - this.#offset;
  }

  public instruction(): {
    readonly opcode: number;
    readonly optional: boolean;
    readonly end: number;
  } {
    const start = this.#offset;
    if (start % PROTOCOL_ALIGNMENT !== 0) fail("picture instruction is not aligned");
    this.require(INSTRUCTION_HEADER_BYTES);
    const opcode = this.u8();
    const flags = this.u8();
    if ((flags & ~INSTRUCTION_FLAG_MASK) !== 0) fail("unsupported picture instruction flags");
    const words = this.u16();
    const length = words === INSTRUCTION_LENGTH_ESCAPE ? this.u32() : words * PROTOCOL_ALIGNMENT;
    const end = checkedAdd(start, length);
    if (length < INSTRUCTION_HEADER_BYTES || length % PROTOCOL_ALIGNMENT !== 0)
      fail("picture instruction length is invalid");
    if (end > this.#input.byteLength) fail("picture instruction runs past the stream");
    return { opcode, optional: (flags & INSTRUCTION_FLAG_OPTIONAL) !== 0, end };
  }

  public seekTo(offset: number): void {
    if (offset < this.#offset || offset > this.#input.byteLength) fail("invalid picture skip");
    this.#offset = offset;
  }

  public u8(): number {
    this.require(1);
    return this.#view.getUint8(this.#offset++);
  }
  public u16(): number {
    this.require(2);
    const value = this.#view.getUint16(this.#offset, true);
    this.#offset += 2;
    return value;
  }
  public u32(): number {
    this.require(4);
    const value = this.#view.getUint32(this.#offset, true);
    this.#offset += 4;
    return value;
  }
  public bytes(length: number): Uint8Array {
    this.require(length);
    const value = this.#input.subarray(this.#offset, this.#offset + length);
    this.#offset += length;
    return value;
  }
  private require(length: number): void {
    if (!Number.isSafeInteger(length) || length < 0 || length > this.remaining)
      fail("truncated picture-resource batch");
  }
}

class Writer {
  readonly #bytes: number[] = [];
  #instructionStart: number | undefined;

  public instruction(opcode: PictureResourceOpcode): void {
    this.closeInstruction();
    this.#instructionStart = this.#bytes.length;
    this.u8(opcode);
    this.u8(0);
    this.u16(0);
  }
  public closeInstruction(): void {
    const start = this.#instructionStart;
    if (start === undefined) return;
    this.#instructionStart = undefined;
    const length = this.#bytes.length - start;
    const words = length / PROTOCOL_ALIGNMENT;
    if (words < INSTRUCTION_LENGTH_ESCAPE) {
      this.#bytes[start + 2] = words & 0xff;
      this.#bytes[start + 3] = (words >>> 8) & 0xff;
      return;
    }
    this.#bytes[start + 2] = INSTRUCTION_LENGTH_ESCAPE & 0xff;
    this.#bytes[start + 3] = (INSTRUCTION_LENGTH_ESCAPE >>> 8) & 0xff;
    const total = checkedAdd(length, PROTOCOL_ALIGNMENT);
    this.#bytes.splice(
      start + INSTRUCTION_HEADER_BYTES,
      0,
      total & 0xff,
      (total >>> 8) & 0xff,
      (total >>> 16) & 0xff,
      (total >>> 24) & 0xff,
    );
  }
  public u8(value: number): void {
    this.#bytes.push(value & 0xff);
  }
  public u16(value: number): void {
    this.#bytes.push(value & 0xff, (value >>> 8) & 0xff);
  }
  public u32(value: number): void {
    if (!Number.isInteger(value) || value < 0 || value > 0xffff_ffff) fail("value is outside u32");
    this.#bytes.push(
      value & 0xff,
      (value >>> 8) & 0xff,
      (value >>> 16) & 0xff,
      (value >>> 24) & 0xff,
    );
  }
  public bytes(value: Uint8Array): void {
    this.#bytes.push(...value);
  }
  public pad(): void {
    while (this.#bytes.length % PROTOCOL_ALIGNMENT !== 0) this.#bytes.push(0);
  }
  public finish(): Uint8Array {
    this.closeInstruction();
    return Uint8Array.from(this.#bytes);
  }
}

function nonzeroId(value: number): number {
  if (!Number.isInteger(value) || value <= 0 || value > 0xffff_ffff)
    fail("picture id must be a non-zero u32");
  return value;
}

function checkedAdd(left: number, right: number): number {
  const value = left + right;
  if (!Number.isSafeInteger(value) || value > 0xffff_ffff) fail("picture byte arithmetic overflow");
  return value;
}

function isKnownOpcode<T extends number>(
  values: Record<string, string | T>,
  value: number,
): value is T {
  return Object.prototype.hasOwnProperty.call(values, value);
}

function fail(message: string): never {
  throw new PictureResourceError(message);
}

File: core/pingo-abi/src/picture_resources.rs
use std::{collections::HashSet, sync::Arc};

use crate::codec::{
    Reader, Writer, checked_padding, finish_instruction, read_header, read_instruction_header,
    validate_encode_instruction_count,
};
use crate::{
    AbiError, DisplayList, MAX_PICTURE_RESIDENT_BYTES, MAX_PICTURE_RESOURCE_INSTRUCTIONS,
    MAX_PICTURE_RESOURCES_BYTES, MAX_RESOURCE_BYTES, PICTURE_RESOURCES_MAGIC,
    PictureResourceOpcode, StreamKind,
};

/// One immutable Picture-resource lifecycle operation.
#[derive(Clone, Debug, PartialEq)]
pub enum PictureResourceCommand {
    /// Publishes a complete DisplayList before a frame may reference it.
    Define {
        /// Session-unique, non-zero Picture generation token.
        picture_id: u32,
        /// Canonical, independently validated DisplayList bytes.
        bytes: Arc<[u8]>,
    },
    /// Releases a Picture after no subsequently replayed frame references it.
    Release {
        /// Previously published Picture generation token.
        picture_id: u32,
    },
}

/// A Picture-resource command plus versioned instruction flags.
#[derive(Clone, Debug, PartialEq)]
pub struct PictureResourceInstruction {
    /// Instruction flags. ABI v17 requires zero.
    pub flags: u8,
    /// Validated lifecycle operation.
    pub command: PictureResourceCommand,
}

/// Transactional Core-to-backend Picture resource delta batch.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct PictureResourceBatch {
    /// Definitions followed by releases for one committed frame.
    pub instructions: Vec<PictureResourceInstruction>,
}

impl PictureResourceBatch {
    /// Fully decodes and validates a Picture resource batch.
    pub fn decode(bytes: &[u8]) -> Result<Self, AbiError> {
        Self::decode_with_report(bytes).map(|(batch, _)| batch)
    }

    /// Decodes while reporting compatible optional instructions.
    pub fn decode_with_report(bytes: &[u8]) -> Result<(Self, crate::DecodeReport), AbiError> {
        let mut reader = Reader::new(bytes);
        let stream = read_header(
            &mut reader,
            PICTURE_RESOURCES_MAGIC,
            MAX_PICTURE_RESOURCES_BYTES,
        )?;
        if stream.declared_count > MAX_PICTURE_RESOURCE_INSTRUCTIONS {
            return Err(AbiError::InstructionCountTooLarge {
                declared: stream.declared_count,
                maximum: MAX_PICTURE_RESOURCE_INSTRUCTIONS,
            });
        }
        let maximum_count =
            u32::try_from(reader.remaining() / 8).map_err(|_| AbiError::ArithmeticOverflow)?;
        if stream.declared_count > maximum_count {
            return Err(AbiError::InstructionCountTooLarge {
                declared: stream.declared_count,
                maximum: maximum_count,
            });
        }
        let capacity =
            usize::try_from(stream.declared_count).map_err(|_| AbiError::ArithmeticOverflow)?;
        let mut instructions = Vec::with_capacity(capacity);
        let mut seen = HashSet::with_capacity(capacity);
        let mut actual_count = 0_u32;
        let mut skipped = 0_u32;
        let mut payload_total = 0_usize;
        while reader.remaining() != 0 {
            let header = read_instruction_header(&mut reader)?;
            let (offset, raw_opcode, flags) = (header.offset, header.opcode, header.flags);
            let Some(opcode) = PictureResourceOpcode::from_u8(raw_opcode) else {
                if header.optional() {
                    skipped = skipped.saturating_add(1);
                    actual_count = actual_count
                        .checked_add(1)
                        .ok_or(AbiError::ArithmeticOverflow)?;
                    reader.seek_to(header.end)?;
                    continue;
                }
                return Err(AbiError::UnknownOpcode {
                    stream: StreamKind::PictureResources,
                    opcode: raw_opcode,
                    offset,
                });
            };
            actual_count = actual_count
                .checked_add(1)
                .ok_or(AbiError::ArithmeticOverflow)?;
            let picture_id = nonzero_id(reader.read_u32()?)?;
            if !seen.insert(picture_id) {
                return Err(AbiError::InvalidValue(
                    "picture id occurs more than once in a batch",
                ));
            }
            let command = match opcode {
                PictureResourceOpcode::DefinePicture => {
                    let payload_bytes = usize::try_from(reader.read_u32()?)
                        .map_err(|_| AbiError::ArithmeticOverflow)?;
                    if payload_bytes > MAX_RESOURCE_BYTES {
                        return Err(AbiError::ResourceTooLarge {
                            actual: payload_bytes,
                            maximum: MAX_RESOURCE_BYTES,
                        });
                    }
                    payload_total = payload_total
                        .checked_add(payload_bytes)
                        .ok_or(AbiError::ArithmeticOverflow)?;
                    if payload_total > MAX_PICTURE_RESIDENT_BYTES {
                        return Err(AbiError::ResourceTooLarge {
                            actual: payload_total,
                            maximum: MAX_PICTURE_RESIDENT_BYTES,
                        });
                    }
                    let payload = reader.read_bytes(payload_bytes)?;
                    DisplayList::decode(payload)?;
                    reader.read_zeroes(checked_padding(payload_bytes)?)?;
                    PictureResourceCommand::Define {
                        picture_id,
                        bytes: Arc::from(payload),
                    }
                }
                PictureResourceOpcode::ReleasePicture => {
                    PictureResourceCommand::Release { picture_id }
                }
            };
            validate_instruction_size(opcode, offset, reader.offset())?;
            finish_instruction(&reader, header)?;
            instructions.push(PictureResourceInstruction { flags, command });
        }
        if actual_count != stream.declared_count {
            return Err(AbiError::InstructionCountMismatch {
                declared: stream.declared_count,
                actual: actual_count,
            });
        }
        Ok((
            Self { instructions },
            crate::DecodeReport {
                skipped_instructions: skipped,
                producer_abi_version: stream.producer_version,
            },
        ))
    }

    /// Encodes one canonical, bounded Picture resource transaction.
    pub fn encode(&self) -> Result<Vec<u8>, AbiError> {
        validate_encode_instruction_count(
            self.instructions.len(),
            0,
            MAX_PICTURE_RESOURCE_INSTRUCTIONS,
        )?;
        let mut writer = Writer::new(PICTURE_RESOURCES_MAGIC);
        let mut seen = HashSet::with_capacity(self.instructions.len());
        let mut payload_total = 0_usize;
        for instruction in &self.instructions {
            if instruction.flags != 0 {
                return Err(AbiError::InvalidValue(
                    "picture resource instruction flags must be zero",
                ));
            }
            let picture_id = match &instruction.command {
                PictureResourceCommand::Define { picture_id, .. }
                | PictureResourceCommand::Release { picture_id } => nonzero_id(*picture_id)?,
            };
            if !seen.insert(picture_id) {
                return Err(AbiError::InvalidValue(
                    "picture id occurs more than once in a batch",
                ));
            }
            match &instruction.command {
                PictureResourceCommand::Define { bytes, .. } => {
                    DisplayList::decode(bytes)?;
                    if bytes.len() > MAX_RESOURCE_BYTES {
                        return Err(AbiError::ResourceTooLarge {
                            actual: bytes.len(),
                            maximum: MAX_RESOURCE_BYTES,
                        });
                    }
                    payload_total = payload_total
                        .checked_add(bytes.len())
                        .ok_or(AbiError::ArithmeticOverflow)?;
                    if payload_total > MAX_PICTURE_RESIDENT_BYTES {
                        return Err(AbiError::ResourceTooLarge {
                            actual: payload_total,
                            maximum: MAX_PICTURE_RESIDENT_BYTES,
                        });
                    }
                    writer.instruction(PictureResourceOpcode::DefinePicture as u8, 0);
                    writer.u32(picture_id);
                    writer
                        .u32(u32::try_from(bytes.len()).map_err(|_| AbiError::ArithmeticOverflow)?);
                    writer.bytes(bytes);
                    writer.pad();
                }
                PictureResourceCommand::Release { .. } => {
                    writer.instruction(PictureResourceOpcode::ReleasePicture as u8, 0);
                    writer.u32(picture_id);
                }
            }
        }
        writer.finish(MAX_PICTURE_RESOURCES_BYTES)
    }

    /// Encodes Core-authored commands whose IDs and nested DisplayLists were
    /// created by the typed paint builder in this process.
    ///
    /// The public/general encoder above remains the validation oracle for
    /// fixtures and external producers. This narrower path deliberately avoids
    /// decoding every just-encoded DisplayList a second time on the product
    /// WASM boundary; the backend decoder still treats the resulting bytes as
    /// untrusted input before installation.
    #[doc(hidden)]
    pub fn encode_core_owned(&self) -> Result<Vec<u8>, AbiError> {
        validate_encode_instruction_count(
            self.instructions.len(),
            0,
            MAX_PICTURE_RESOURCE_INSTRUCTIONS,
        )?;
        let mut writer = Writer::new(PICTURE_RESOURCES_MAGIC);
        let mut payload_total = 0_usize;
        for instruction in &self.instructions {
            if instruction.flags != 0 {
                return Err(AbiError::InvalidValue(
                    "picture resource instruction flags must be zero",
                ));
            }
            let picture_id = match &instruction.command {
                PictureResourceCommand::Define { picture_id, .. }
                | PictureResourceCommand::Release { picture_id } => nonzero_id(*picture_id)?,
            };
            match &instruction.command {
                PictureResourceCommand::Define { bytes, .. } => {
                    if bytes.len() > MAX_RESOURCE_BYTES {
                        return Err(AbiError::ResourceTooLarge {
                            actual: bytes.len(),
                            maximum: MAX_RESOURCE_BYTES,
                        });
                    }
                    payload_total = payload_total
                        .checked_add(bytes.len())
                        .ok_or(AbiError::ArithmeticOverflow)?;
                    if payload_total > MAX_PICTURE_RESIDENT_BYTES {
                        return Err(AbiError::ResourceTooLarge {
                            actual: payload_total,
                            maximum: MAX_PICTURE_RESIDENT_BYTES,
                        });
                    }
                    writer.instruction(PictureResourceOpcode::DefinePicture as u8, 0);
                    writer.u32(picture_id);
                    writer
                        .u32(u32::try_from(bytes.len()).map_err(|_| AbiError::ArithmeticOverflow)?);
                    writer.bytes(bytes);
                    writer.pad();
                }
                PictureResourceCommand::Release { .. } => {
                    writer.instruction(PictureResourceOpcode::ReleasePicture as u8, 0);
                    writer.u32(picture_id);
                }
            }
        }
        writer.finish(MAX_PICTURE_RESOURCES_BYTES)
    }
}

fn nonzero_id(value: u32) -> Result<u32, AbiError> {
    if value == 0 {
        Err(AbiError::InvalidValue("picture id must be non-zero"))
    } else {
        Ok(value)
    }
}

fn validate_instruction_size(
    opcode: PictureResourceOpcode,
    start: usize,
    end: usize,
) -> Result<(), AbiError> {
    let actual = end.checked_sub(start).ok_or(AbiError::ArithmeticOverflow)?;
    if let Some(expected) = opcode.fixed_bytes()
        && actual != expected
    {
        return Err(AbiError::InstructionLengthMismatch {
            opcode: opcode as u8,
            offset: start,
            expected,
            actual,
        });
    }
    if actual < opcode.minimum_bytes() {
        return Err(AbiError::InstructionLengthMismatch {
            opcode: opcode as u8,
            offset: start,
            expected: opcode.minimum_bytes(),
            actual,
        });
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{DisplayCommand, DisplayInstruction};

    fn list() -> Arc<[u8]> {
        Arc::from(
            DisplayList {
                instructions: vec![
                    DisplayInstruction {
                        flags: 0,
                        command: DisplayCommand::Save,
                    },
                    DisplayInstruction {
                        flags: 0,
                        command: DisplayCommand::Restore,
                    },
                ],
            }
            .encode()
            .expect("display list"),
        )
    }

    #[test]
    fn round_trips_define_and_release() {
        let batch = PictureResourceBatch {
            instructions: vec![
                PictureResourceInstruction {
                    flags: 0,
                    command: PictureResourceCommand::Define {
                        picture_id: 7,
                        bytes: list(),
                    },
                },
                PictureResourceInstruction {
                    flags: 0,
                    command: PictureResourceCommand::Release { picture_id: 8 },
                },
            ],
        };
        let encoded = batch.encode().expect("encode");
        assert_eq!(encoded, batch.encode_core_owned().expect("Core encode"));
        assert_eq!(
            PictureResourceBatch::decode(&encoded).expect("decode"),
            batch
        );
    }

    #[test]
    fn rejects_duplicate_ids_and_malformed_nested_lists() {
        let duplicate = PictureResourceBatch {
            instructions: vec![
                PictureResourceInstruction {
                    flags: 0,
                    command: PictureResourceCommand::Release { picture_id: 4 },
                },
                PictureResourceInstruction {
                    flags: 0,
                    command: PictureResourceCommand::Release { picture_id: 4 },
                },
            ],
        };
        assert!(duplicate.encode().is_err());
        let malformed = PictureResourceBatch {
            instructions: vec![PictureResourceInstruction {
                flags: 0,
                command: PictureResourceCommand::Define {
                    picture_id: 9,
                    bytes: Arc::from([1_u8, 2, 3, 4]),
                },
            }],
        };
        assert!(malformed.encode().is_err());
    }
}
